src/error-demo/error-demo.cc

```diff
@@ -1,66 +1,157 @@
-#include "error.hh"
+#include "logging.hh"
 #include "nixexpr.hh"
+#include "util.hh"
 
 #include <iostream>
 #include <optional>
 
+using namespace nix;
+
+MakeError(DemoError, Error);
+
 int main()
 {
-    using namespace nix;
+    verbosity = lvlVomit;
 
     // In each program where errors occur, this has to be set.
     ErrorInfo::programName = std::optional("error-demo");
 
+    // 'DemoError' appears as error name.
```

Comment:
I wonder if it would be possible to turn this `demo` file into unit tests instead? Would that be feasible? WDYT @bburdette ?

Reply:
I am not opposed.  What do you envision as tests?  String comparison of predicted vs actual output?

Reply:
Yes i think that is what I would envision. I have to admit i didn’t yet find the time to take a closer look so maybe some parts might be hard(er) to test when log functions just print to stderr bur whenever an Exception is thrown with a message that encodes information in a certain format that could be tested.

Reply:
I actually just spent some time on this and am almost done converting everything from your `error-demo.cc` into a unit test. I'll probably have it done later this evening for you to check out.

Reply:
Wow just saw this!  I've been working on it too, but u

Reply:
Please have a look at:
https://github.com/gilligan/nix/commit/952e72c804ce79ffeee8f58261d4cbdeb9d56ac6

I would suggest to remove `error-demo` and replace it with that. Some notes:

- Shortly after i started i realized that this would really be tests for `logging.h` 
- I am using the google test to capture output and assert that. That seemed like the best approach to me
- You will see notes with `XXX` where i don't quite understand why `description` is not set.
- The asserts do look a bit ugly. It might be nicer to use c++ raw strings so the expected output is actually human readable. Using those in the `json.hh` tests:

```c++
        ASSERT_EQ(out.str(),
R"#({
  "list": [
    "element"
  ]
})#");
```

I was using those at first but then realized that because of the escape codes for coloring they wouldn't work. I guess ideally we'd still use those and add the escape codes (?). Or we leave that as an improvement for later. I'll leave it up to you.

Reply:
Nice!  The testing::internal::CaptureStderr() is helpful.  My approach was doing a test in bash; this is better I think.  

I do wonder about the empty strings on Talkative and etc.  On the plus side, its only the logError and logWarning that are used in actual code at this point.  Seems like a problem with the stderr capture somehow?  "Chatty" still has a "talkative" description; I made a note.  

Anyway overall LGTM.  Thanks for expediting this!